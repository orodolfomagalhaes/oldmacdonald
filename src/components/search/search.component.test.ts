import { expect, it, afterEach, beforeEach, describe, vi } from 'vitest';
import {SearchComponent} from "./search.component.ts";

describe("SearchComponent", () => {
  let component: any;

  const mockTooltip = vi.fn();
  // @ts-ignore
  globalThis.bootstrap = {
    Tooltip: mockTooltip
  }

  beforeEach(() => {
    component = new SearchComponent();
    document.body.appendChild(component);
  });

  afterEach(() => {
    document.body.innerHTML = ``;
  });

  it('should create the search component', () => {
    expect(component).toBeTruthy();
    expect(component).toMatchSnapshot();
  });

  it('should listen to the search input event', () => {
    const eventListener = vi.fn();
    const input: any = document.querySelector('#search-input');
    input.addEventListener('input', eventListener);

    input.value = 'qwerty';
    const event = new Event('input');
    input.dispatchEvent(event);

    expect(eventListener).toHaveBeenCalledTimes(1);
    expect(eventListener.mock.calls[0][0].target.value).toBe('qwerty');
    input.removeEventListener('input', eventListener);
  });

  it('should show clear button when input has value', () => {
    const spyShowClearButton = vi.spyOn(component, 'showClearButton');
    const input: any = document.querySelector('#search-input');
    input.value = 'qwerty';
    const event = new Event('input', { bubbles: true });
    input.dispatchEvent(event);

    const clearButton = document.querySelector('.btn-clear');
    expect(spyShowClearButton).toHaveBeenCalledTimes(1);
    expect(clearButton?.classList.contains('d-none')).toBeFalsy();
  });

  it('should hide clear button when input has no value', () => {
    const input: any = document.querySelector('#search-input');
    const clearButton = document.querySelector('.btn-clear');
    // show
    const spyShowClearButton = vi.spyOn(component, 'showClearButton');
    input.value = 'qwerty';
    const event = new Event('input', { bubbles: true });
    input.dispatchEvent(event);

    expect(spyShowClearButton).toHaveBeenCalledTimes(1);
    expect(clearButton?.classList.contains('d-none')).toBeFalsy();

    // hide
    const spyHideClearButton = vi.spyOn(component, 'hideClearButton');
    input.value = '';
    input.dispatchEvent(event);

    expect(spyHideClearButton).toHaveBeenCalledTimes(1);
    expect(clearButton?.classList.contains('d-none')).toBeTruthy();
  });

  it('should listen to the clear button click', () => {
    const eventListener = vi.fn();
    const button: any = document.querySelector('.btn-clear');
    button.addEventListener('click', eventListener);
    button?.click();

    expect(eventListener).toHaveBeenCalledTimes(1);
    button?.removeEventListener('click', eventListener);
  });

  it('should dispatch search-event event when clear button is clicked', () => {
    vi.useFakeTimers();
    const spyDebounce = vi.spyOn(component, 'debounce');
    const spyDispatchSearchEvent = vi.spyOn(component, 'dispatchSearchEvent');

    const input: any = document.querySelector('#search-input');
    input.value = 'qwerty';
    const event = new Event('input', { bubbles: true });
    input.dispatchEvent(event);

    const button: any = document.querySelector('.btn-clear');
    button?.click();

    expect(spyDebounce).toHaveBeenCalledTimes(2);
    vi.advanceTimersByTime(700);
    expect(spyDispatchSearchEvent).toHaveBeenCalledTimes(1);
  });
});