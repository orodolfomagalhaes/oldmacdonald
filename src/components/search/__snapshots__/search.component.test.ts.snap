// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchComponent > should create the search component 1`] = `
<search-component>
  
    
  <div
    class="search-container mb-3"
  >
    
      
    <div
      class="d-flex align-items-center gap-2 search-input-group"
    >
      
        
      <div
        class="input-group"
      >
        
          
        <i
          aria-hidden="true"
          class="fa fa-search input-group-text"
          id="basic-addon1"
        />
        
          
        <input
          aria-describedby="basic-addon1"
          aria-label="Search farm by full name"
          class="form-control"
          id="search-input"
          placeholder="Farm name"
          type="text"
        />
        
        
      </div>
      
        
      <button
        class="btn btn-link btn-clear d-none"
        type="button"
      >
        clear
      </button>
      
      
    </div>
    
      
    <i
      aria-hidden="true"
      aria-label="Please search using the full farm name, as the current version of json-server does not support filtering with regular expressions."
      class="fa fa-info-circle fa-1"
      data-bs-placement="right"
      data-bs-title="Please search using the full farm name, as the current version of json-server does not support filtering with regular expressions."
      data-bs-toggle="tooltip"
    />
         
    
  </div>
</search-component>
`;
